compiler/generator_rust.c

```diff
@@ -0,0 +1,1369 @@
+/* (i.e. upper-case for globals, snake case for fields etc.) */
+/* To allow warning free compilation of generated code and */
+/* consistency with snowball variable namings we allow some kind of warnings here */
+static void generate_allow_warnings(struct generator * g) {
+
+    w(g, "#![allow(non_upper_case_globals)]~N");
+    w(g, "#![allow(non_snake_case)]~N");
+    w(g, "#![allow(unused_variables)]~N");
+    w(g, "#![allow(unused_mut)]~N");
+}
+
+static void generate_class_begin(struct generator * g) {
+
+    w(g, "use snowball::SnowballEnv;~N");
+    w(g, "use snowball::Among;~N~N");  
+}
+
+static void generate_among_table(struct generator * g, struct among * x) {
+
+    struct amongvec * v = x->b;
+
+    g->I[0] = x->number;
+    g->I[1] = x->literalstring_count;
+
+    w(g, "~Mstatic A_~I0: &'static [Among<Context>; ~I1] = &[~N~+");
+    {
+        int i;
+        for (i = 0; i < x->literalstring_count; i++)
+        {
+            g->I[0] = i;
+            g->I[1] = v->i;
+            g->I[2] = v->result;
+            g->L[0] = v->b;
+            g->S[0] = ",";
+
+            w(g, "~MAmong(~L0, ~I1, ~I2, ");
+            if (v->function != 0)
+            {
+                w(g, "Some(&");
+                write_varname(g, v->function);
+                w(g, ")");
+            } else {
+              w(g, "None");
+            }
+            w(g, ")~S0~N");
+            v++;
+        }
+    }
+    w(g, "~-~M];~N~N");
+}
+
+static void generate_amongs(struct generator * g) {
+
+    struct among * x = g->analyser->amongs;
+    while (x != 0) {
+        generate_among_table(g, x);
+        x = x->next;
+    }
+}
+
+static void set_bit(symbol * b, int i) { b[i/8] |= 1 << i%8; }
+
+static void generate_grouping_table(struct generator * g, struct grouping * q) {
+
+    int range = q->largest_ch - q->smallest_ch + 1;
+    int size = (range + 7)/ 8;  /* assume 8 bits per symbol */
+    symbol * b = q->b;
+    symbol * map = create_b(size);
+    int i;
+    for (i = 0; i < size; i++) map[i] = 0;
+
+    for (i = 0; i < SIZE(b); i++) set_bit(map, b[i] - q->smallest_ch);
+
+    g->V[0] = q->name;
+    g->I[0] = size;
+    w(g, "~Mstatic ~W0: &'static [u8; ~I0] = &[");
+    for (i = 0; i < size; i++) {
+        write_int(g, map[i]);
+        if (i < size - 1) w(g, ", ");
+    }
+    w(g, "];~N~N");
+    lose_b(map);
+}
+
+static void generate_groupings(struct generator * g) {
+    struct grouping * q;
+    for (q = g->analyser->groupings; q; q = q->next) {
+        generate_grouping_table(g, q);
+    }
+}
+
+
+static void generate_members(struct generator * g) {
+  
+    struct name * q;
+    w(g, "#[derive(Clone)]~N");
+    w(g, "struct Context {~+~N");
+    for (q = g->analyser->names; q; q = q->next) {
+        g->V[0] = q;
+        switch (q->type) {
+            case t_string:
+                w(g, "~M~W0: String,~N");
+                break;
+            case t_integer:
+                w(g, "~M~W0: usize,~N");
```

Comment:
@JDemler Sorry, I failed to spot this before but the Rust backend maps Snowball integers to an unsigned type, but [the manual](http://snowballstem.org/compiler/snowman.html) clearly says that they should be "signed integers".  I think it's probably simplest to change this globally (so the cursor and limits become signed too) because of commands like `atlimit`, `tolimit`, the ability to use `cursor` in numeric expressions, etc but I'm not sure if there are consequences to making this change as I've never programmed in Rust.  Also not sure what Rust type is most appropriate - I guess `i32` or `isize` from looking at https://doc.rust-lang.org/book/second-edition/ch03-02-data-types.html (in practice Snowball strings are at most a few hundred characters so a 32-bit limit is not problematic).

Reply:
Let me see if I understand the problem correctly:

1. Using AEs one could possibly create negative integer values
2.  `minint` is defined as the minimum negative number

The rust backend is not able to handle both cases corretly.
1. first would probably result in a runtime panic.
2. `minint` returns `0` in generated code

Beyond AEs though there is no chance to ever encounter a negative number, as they are filtered:
``` rust
   let c = env.byte_index_for_hop(3);
    if 0 as i32 > c || c > env.limit as i32 {
        return false;
    }
    env.cursor = c as usize;
```

Do I understand the problem correctly?
Anyways... the simplest solution would probably be to change everything to `isize`.
Not only the local variables, but also `cursor`, `limit`, `limit_backward`, `bra` and `ket` inside `SnowballEnv`

Are there any real live examples of AEs resulting in negative values?

Reply:
I think your summary is accurate.

You can also set an integer variable without involving an AE by using `setmark x` but that should never give a negative value, so I think any negative value must come from an AE.

> Are there any real live examples of AEs resulting in negative values?

I noticed it while looking at the code in `generator_rust.c` (I'm working on a compiler optimisation to "localise" variables - i.e. map a Snowball variable to a function-local variable in the target language when appropriate).

I don't think there aren't any examples in the current stemmers we ship.  There could be cases in Snowball code that's not been submitted to us, but realistically it's probably a theoretical issue at the moment.  However, I think it is worth addressing this as negative integers are something one might reasonably want to use - we're not in the domain of weird programs written only to prove a pedantic point.

Reply:
All right. I'll try to whip up a PR next week.